robot_skills/src/robot_skills/functionalities/arm/guarded_motion.py

```diff
@@ -11,55 +11,81 @@ class GuardedMotionFunc(RobotFunc):
+                                                {
+                                                    "move_down_until_force_sensor_edge_up": move_down_until_force_sensor_edge_up,
+                                                    "_create_lower_force_sensing_goal": _create_lower_force_sensing_goal,
+                                                    "_create_retract_force_sensing_goal": _create_retract_force_sensing_goal
+                                                })
 
     def check_requirements(self, arm):
-        "Check that the arm has at least one force sensor"
+        """
+        Check that the arm has at least one force sensor
+        """
         for part in arm.parts:
             if isinstance(arm.parts[part], ForceSensor):
                 return True
         return False
 
 
-def move_down_until_force_sensor_edge_up(self, force_sensor=None, timeout=10, retract_distance=0.01):
+def move_down_until_force_sensor_edge_up(self, force_sensor=None, timeout=10, retract_distance=0.01,
+                                         distance_move_down=None):
     """
-    Move down the arm (hero specific, only joint arm_lift_joint) until the force sensor detects an edge up
+    Move down the arm (hero specific, only joint arm_lift_joint) until one of 3 things:
+        - Force sensor detects an edge up
+        - Timeout
+        - Force sensor has moved down the distance specified
 
     A 'force_sensor.TimeOutException' will be raised if no edge up is detected within timeout
 
+    :param force_sensor: ForceSensor of the arm
     :param timeout: Max duration for edge up detection
     :param retract_distance: How much to retract if we have reached a surface
+    :param distance_move_down: How much to move down until we reach a surface
     """
-    if not force_sensor:
+    if force_sensor is None:
         force_sensor = self.parts["force_sensor"]
 
     # Fill with required joint names (desired in hardware / gazebo impl)
-    goal = self._create_lower_force_sensing_goal(timeout)
+    goal = self._create_lower_force_sensing_goal(distance_move_down, timeout)
     self._ac_joint_traj.send_goal(goal)
 
     force_sensor.wait_for_edge_up(timeout)
     self.cancel_goals()
 
-    goal = self._create_retract_force_sensing_goal(0.5, timeout)
+    goal = self._create_retract_force_sensing_goal(retract_distance, timeout)
     self._ac_joint_traj.send_goal(goal)
 
 
-def _create_lower_force_sensing_goal(self, timeout):
+def _create_lower_force_sensing_goal(self, distance_move_down, timeout):
+    # Gets current joint state
     current_joint_state = self.get_joint_states()
-    current_joint_state['arm_lift_joint'] = 0   # TODO make this function not HERO-specific
+
+    # Sets the joint state to move down a certain distance (or 0 if distance is not set)
+    if distance_move_down is not None:
+        current_joint_state['arm_lift_joint'] = max(0, current_joint_state['arm_lift_joint'] - distance_move_down)
```

Comment:
0 should be the lower limit of the joint.

Reply:
0 is the lowest the arm lift joint state value can be. With the distance_move_down it can be greater than 0